File: valid-palindrome-ii/Solution680.java
// https://leetcode.com/problems/valid-palindrome-ii/

class Solution680 {

    // T: O(n), S:(1)
    public boolean validPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return palindromeHelper(s, left, right - 1) ||
                        palindromeHelper(s, left + 1, right);
            }
            left++;
            right--;
        }
        return true;
    }

    private boolean palindromeHelper(String s, int i, int j) {
        while (i < j) {
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }

    public static void main(String[] args) {
        var sol = new Solution680();
        var s1 = "aba";
        var s2 = "abca";
        var s3 = "abc";
        System.out.println(sol.validPalindrome(s1)); // true
        System.out.println(sol.validPalindrome(s2)); // true
        System.out.println(sol.validPalindrome(s3)); // false
    }

}

File: valid-palindrome-ii/solution680.py
# https://leetcode.com/problems/valid-palindrome-ii/

class Solution680:

    # T: O(n), S:(1)
    def validPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        while left < right:
            if s[left] != s[right]:
                return self.palindromeHelper(s, left, right - 1) or self.palindromeHelper(s, left + 1, right)
            left += 1
            right -= 1
        return True

    def palindromeHelper(self, s: str, i: int, j: int) -> bool:
        while i < j:
            if s[i] != s[j]:
                return False
            i += 1
            j -= 1
        return True


def main() -> None:
    sol = Solution680()
    s1 = "aba"
    s2 = "abca"
    s3 = "abc"
    print(sol.validPalindrome(s1))  # True
    print(sol.validPalindrome(s2))  # True
    print(sol.validPalindrome(s3))  # False


if __name__ == "__main__":
    main()
